### PR TITLE
Optional otLinkRaw Features

### PR DESCRIPTION
**Pull request by nibanks**
When a Host uses the Spinel interface to control the `otLinkRaw*` interface it practically requires the device to support of all the `otRadioCaps`. Many of these features are not supported in hardware by our current example platforms. This change adds support for each feature in software, controllable by compile-time flags. All of this is still gated on the compile time flag that controls the `otLinkRaw*` API.

In summary, with this change, all example platforms (excluding posix) should fully support a Host using the `otLinkRaw*` APIs via Spinel.

### CODE REVIEW

**Comment by nibanks** [comment]
I'm not sure why appveyor is having issues, but the change build fine on [my branch](https://ci.appveyor.com/project/nibanks/openthread/build/0.1.625).

**Comment by abtink** [comment]
Happened to another PR as well... re-pushing the commit  kicked the build and fixed it for me...

File: src/core/api/link_raw.hpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+#if OPENTHREAD_ENABLE_SOFTWARE_ACK_TIMEOUT || OPENTHREAD_ENABLE_SOFTWARE_RETRANSMIT || OPENTHREAD_ENABLE_SOFTWARE_ENERGY_SCAN
+#define OPENTHREAD_LINKRAW_TIMER_REQUIRED 1
+#else
+#define OPENTHREAD_LINKRAW_TIMER_REQUIRED 0
+#endif
+
+namespace Thread {
+
+class LinkRaw
```
Comment (abtink) [inline src/core/api/link_raw.hpp]: :+1:

File: configure.ac [hunk 2]
```diff
@@ -872,6 +872,96 @@ AM_CONDITIONAL([OPENTHREAD_ENABLE_RAW_LINK_API], [test "${enable_raw_link_api}"
 AC_DEFINE_UNQUOTED([OPENTHREAD_ENABLE_RAW_LINK_API],[${OPENTHREAD_ENABLE_RAW_LINK_API}],[Define to 1 if you want to enable raw link-layer API])
 
 #
+# Software ACK Timeout Logic
+#
+
+AC_ARG_ENABLE(sw_ack_timeout,
```
Comment (abtink) [inline configure.ac]: This is fine as is but I think these may be more suited as config option defines in `openthread-core-default-config.h`. 
I think enabling `raw apis` is a good high-level feature suited for `configure.ac` but the capabilities may be better in the config header file. Thoughts?

Comment (nibanks) [inline configure.ac]: The reason I put them in here is because it made configuring the makefiles for each platform nice, clean and explicit, in my opinion.

Comment (abtink) [inline configure.ac]: I agree the makefiles are neat and clean with this approach.  But I am afraid we may need to add many new config features in `configure.ac`...
 
I think in long-term we will probably need a separate config-file for each platform (we will need it if we want to change the number/size of different buffers, etc... which are platform dependant )... And we already have it for nrf52840.  We can start adding it now for each platform. 

Thoughts?

Comment (pvanhorn) [inline configure.ac]: There is something to be said for a developer to be able to successfully navigate the configure options OpenThread makes available.  Towards that end, I agree with Abtin that these options rightfully belong in the openthread-core-default-config.h.  If i was new to Openthread and had to guess where these options might exist, that file would be my first guess.

Comment (nibanks) [inline configure.ac]: So your preference would be that I go and add a platform specific core config header (like nrf52840) for each platform?

Comment (pvanhorn) [inline configure.ac]: Yes, that is what we do and it works great.

**Comment by nibanks** [comment]
@abtink @pvanhorn Are the most recent changes what you wanted? Do you have any other issues with this PR?

File: src/core/openthread-core-default-config.h [hunk 3]
```diff
@@ -559,4 +559,34 @@
 #define OPENTHREAD_CONFIG_PLATFORM_ASSERT_MANAGEMENT            0
 #endif
 
+/**
```
Comment (pvanhorn) [inline src/core/openthread-core-default-config.h]: To keep with convention of this file, these new options should be named OPENTHREAD_CONFIG_...

**Comment by pvanhorn** [comment]
aside from the one style comment I just posted, I believe your changes satisfy my only concern.

File: examples/Makefile-cc2538 [hunk 4]
```diff
@@ -70,11 +70,19 @@ ifeq ($(DHCP6_CLIENT),1)
...
+CONFIG_FILE      = OPENTHREAD_PROJECT_CORE_CONFIG_FILE='\"openthread-core-cc2538-config.h\"'
+CONFIG_FILE_PATH = $(AbsTopSourceDir)/examples/platforms/cc2538/
+
 COMMONCFLAGS                   := \
     -fdata-sections               \
     -ffunction-sections           \
     -Os                           \
     -g                            \
+    -D$(CONFIG_FILE)              \
+    -I$(CONFIG_FILE_PATH)         \
```
Comment (nibanks) [inline examples/Makefile-cc2538]: These changes don't seem to work. I put some bad characters in code that would only compile in if the correct defines from this header were set and I don't get any compile errors. Can anyone tell me what I am doing wrong? I copied this from `Makefile-nrf52840`, but that makefile seem to get the build errors as expected.

Comment (nibanks) [inline examples/Makefile-cc2538]: OK. False alarm. I had a setup issue.

**Comment by pvanhorn** [comment]
LGTM
